{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  6%|▌         | 14/240 [00:14<04:34,  1.22s/it]Process Process-6:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 391, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 2, in raise_from\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 387, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 1198, in getresponse\n",
      "    response.begin()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 297, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 258, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/socket.py\", line 576, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 937, in recv_into\n",
      "    return self.read(nbytes, buffer)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 799, in read\n",
      "    return self._sslobj.read(len, buffer)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 583, in read\n",
      "    v = self._sslobj.read(len, buffer)\n",
      "socket.timeout: The read operation timed out\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/adapters.py\", line 423, in send\n",
      "    timeout=timeout\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 643, in urlopen\n",
      "    _stacktrace=sys.exc_info()[2])\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/util/retry.py\", line 334, in increment\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/packages/six.py\", line 686, in reraise\n",
      "    raise value\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 594, in urlopen\n",
      "    chunked=chunked)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 393, in _make_request\n",
      "    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 313, in _raise_timeout\n",
      "    raise ReadTimeoutError(self, url, \"Read timed out. (read timeout=%s)\" % timeout_value)\n",
      "requests.packages.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='storage.data.gov.sg', port=443): Read timed out. (read timeout=1)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"../../limited-requests/limited_requests.py\", line 33, in _fetch\n",
      "    dataset_tuples = datafy.get(uri)\n",
      "  File \"../../datafy/datafy.py\", line 59, in get\n",
      "    content_length = int(requests.head(uri, timeout=1).headers['content-length'])\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/api.py\", line 96, in head\n",
      "    return request('head', url, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/api.py\", line 56, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/sessions.py\", line 488, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/sessions.py\", line 609, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/adapters.py\", line 499, in send\n",
      "    raise ReadTimeout(e, request=request)\n",
      "requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='storage.data.gov.sg', port=443): Read timed out. (read timeout=1)\n",
      "/home/alex/Desktop/urban-physiology/src/glossarizers/ckan-glossarizer/ckan_glossarizer.py:231: UserWarning: Couldn't parse the URI https://storage.data.gov.sg/age-specific-death-rates-annual/resources/crude-death-rates-by-sex-2017-03-03T16-00-13Z.csv due to a transient network failure.\n",
      "  .format(resource['id']['resource'])\n",
      "  8%|▊         | 20/240 [00:21<03:55,  1.07s/it]Process Process-10:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 391, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 2, in raise_from\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 387, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 1198, in getresponse\n",
      "    response.begin()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 297, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 258, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/socket.py\", line 576, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 937, in recv_into\n",
      "    return self.read(nbytes, buffer)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 799, in read\n",
      "    return self._sslobj.read(len, buffer)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 583, in read\n",
      "    v = self._sslobj.read(len, buffer)\n",
      "socket.timeout: The read operation timed out\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/adapters.py\", line 423, in send\n",
      "    timeout=timeout\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 643, in urlopen\n",
      "    _stacktrace=sys.exc_info()[2])\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/util/retry.py\", line 334, in increment\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/packages/six.py\", line 686, in reraise\n",
      "    raise value\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 594, in urlopen\n",
      "    chunked=chunked)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 393, in _make_request\n",
      "    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 313, in _raise_timeout\n",
      "    raise ReadTimeoutError(self, url, \"Read timed out. (read timeout=%s)\" % timeout_value)\n",
      "requests.packages.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='storage.data.gov.sg', port=443): Read timed out. (read timeout=1)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"../../limited-requests/limited_requests.py\", line 33, in _fetch\n",
      "    dataset_tuples = datafy.get(uri)\n",
      "  File \"../../datafy/datafy.py\", line 59, in get\n",
      "    content_length = int(requests.head(uri, timeout=1).headers['content-length'])\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/api.py\", line 96, in head\n",
      "    return request('head', url, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/api.py\", line 56, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/sessions.py\", line 488, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/sessions.py\", line 609, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/adapters.py\", line 499, in send\n",
      "    raise ReadTimeout(e, request=request)\n",
      "requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='storage.data.gov.sg', port=443): Read timed out. (read timeout=1)\n",
      "/home/alex/Desktop/urban-physiology/src/glossarizers/ckan-glossarizer/ckan_glossarizer.py:231: UserWarning: Couldn't parse the URI https://storage.data.gov.sg/air-accident-investigation-reports-by-date/resources/air-accident-investigation-reports-by-date-2016-08-05T02-54-38Z.csv due to a transient network failure.\n",
      "  .format(resource['id']['resource'])\n",
      "100%|██████████| 240/240 [04:43<00:00,  1.12s/it]\n"
     ]
    }
   ],
   "source": [
    "import ckan_glossarizer; ckan_glossarizer.write_glossary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/2 [00:00<?, ?it/s]Process Process-129:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 391, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 2, in raise_from\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 387, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 1198, in getresponse\n",
      "    response.begin()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 297, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 258, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/socket.py\", line 576, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 937, in recv_into\n",
      "    return self.read(nbytes, buffer)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 799, in read\n",
      "    return self._sslobj.read(len, buffer)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 583, in read\n",
      "    v = self._sslobj.read(len, buffer)\n",
      "socket.timeout: The read operation timed out\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/adapters.py\", line 423, in send\n",
      "    timeout=timeout\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 643, in urlopen\n",
      "    _stacktrace=sys.exc_info()[2])\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/util/retry.py\", line 334, in increment\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/packages/six.py\", line 686, in reraise\n",
      "    raise value\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 594, in urlopen\n",
      "    chunked=chunked)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 393, in _make_request\n",
      "    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 313, in _raise_timeout\n",
      "    raise ReadTimeoutError(self, url, \"Read timed out. (read timeout=%s)\" % timeout_value)\n",
      "requests.packages.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='storage.data.gov.sg', port=443): Read timed out. (read timeout=1)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"../../limited-requests/limited_requests.py\", line 33, in _fetch\n",
      "    dataset_tuples = datafy.get(uri)\n",
      "  File \"../../datafy/datafy.py\", line 59, in get\n",
      "    content_length = int(requests.head(uri, timeout=1).headers['content-length'])\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/api.py\", line 96, in head\n",
      "    return request('head', url, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/api.py\", line 56, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/sessions.py\", line 488, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/sessions.py\", line 609, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/adapters.py\", line 499, in send\n",
      "    raise ReadTimeout(e, request=request)\n",
      "requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='storage.data.gov.sg', port=443): Read timed out. (read timeout=1)\n",
      "/home/alex/Desktop/urban-physiology/src/glossarizers/ckan-glossarizer/ckan_glossarizer.py:231: UserWarning: Couldn't parse the URI https://storage.data.gov.sg/age-specific-death-rates-annual/resources/crude-death-rates-by-sex-2017-03-03T16-00-13Z.csv due to a transient network failure.\n",
      "  .format(resource['id']['resource'])\n",
      " 50%|█████     | 1/2 [00:01<00:01,  1.92s/it]Process Process-130:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 391, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 2, in raise_from\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 387, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 1198, in getresponse\n",
      "    response.begin()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 297, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/http/client.py\", line 258, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/socket.py\", line 576, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 937, in recv_into\n",
      "    return self.read(nbytes, buffer)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 799, in read\n",
      "    return self._sslobj.read(len, buffer)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/ssl.py\", line 583, in read\n",
      "    v = self._sslobj.read(len, buffer)\n",
      "socket.timeout: The read operation timed out\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/adapters.py\", line 423, in send\n",
      "    timeout=timeout\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 643, in urlopen\n",
      "    _stacktrace=sys.exc_info()[2])\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/util/retry.py\", line 334, in increment\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/packages/six.py\", line 686, in reraise\n",
      "    raise value\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 594, in urlopen\n",
      "    chunked=chunked)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 393, in _make_request\n",
      "    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py\", line 313, in _raise_timeout\n",
      "    raise ReadTimeoutError(self, url, \"Read timed out. (read timeout=%s)\" % timeout_value)\n",
      "requests.packages.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='storage.data.gov.sg', port=443): Read timed out. (read timeout=1)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/multiprocessing/process.py\", line 249, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/multiprocessing/process.py\", line 93, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"../../limited-requests/limited_requests.py\", line 33, in _fetch\n",
      "    dataset_tuples = datafy.get(uri)\n",
      "  File \"../../datafy/datafy.py\", line 59, in get\n",
      "    content_length = int(requests.head(uri, timeout=1).headers['content-length'])\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/api.py\", line 96, in head\n",
      "    return request('head', url, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/api.py\", line 56, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/sessions.py\", line 488, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/sessions.py\", line 609, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/home/alex/miniconda3/envs/urban-physiology/lib/python3.5/site-packages/requests/adapters.py\", line 499, in send\n",
      "    raise ReadTimeout(e, request=request)\n",
      "requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='storage.data.gov.sg', port=443): Read timed out. (read timeout=1)\n",
      "/home/alex/Desktop/urban-physiology/src/glossarizers/ckan-glossarizer/ckan_glossarizer.py:231: UserWarning: Couldn't parse the URI https://storage.data.gov.sg/air-accident-investigation-reports-by-date/resources/air-accident-investigation-reports-by-date-2016-08-05T02-54-38Z.csv due to a transient network failure.\n",
      "  .format(resource['id']['resource'])\n",
      "100%|██████████| 2/2 [00:03<00:00,  1.88s/it]\n"
     ]
    }
   ],
   "source": [
    "ckan_glossarizer.write_glossary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      "  0%|          | 0/2 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> /home/alex/Desktop/urban-physiology/src/glossarizers/ckan-glossarizer/ckan_glossarizer.py(218)write_glossary()\n",
      "-> repr = limited_requests.limited_get(resource['id']['resource'], q)\n",
      "(Pdb) l\n",
      "213  \t                succeeded = True\n",
      "214  \t\n",
      "215  \t            # If we error out, this is a packaged/gzipped file. Do sizing the basic way, with a GET request.\n",
      "216  \t            except KeyError:\n",
      "217  \t                import pdb; pdb.set_trace()\n",
      "218  ->\t                repr = limited_requests.limited_get(resource['id']['resource'], q)\n",
      "219  \t                try:\n",
      "220  \t                    glossarized_resource['sizing'] = {\n",
      "221  \t                        'filesize': repr[0]['filesize']\n",
      "222  \t                    }\n",
      "223  \t                    glossarized_resource['dataset'] = {\n",
      "(Pdb) n\n",
      "> /home/alex/Desktop/urban-physiology/src/glossarizers/ckan-glossarizer/ckan_glossarizer.py(219)write_glossary()\n",
      "-> try:\n",
      "(Pdb) repr\n",
      "[{'dataset': '.', 'filesize': 6853, 'resource': 'https://storage.data.gov.sg/age-specific-death-rates-annual/resources/crude-death-rates-by-sex-2017-03-03T16-00-13Z.csv', 'mimetype': 'text/csv', 'extension': 'csv'}]\n",
      "(Pdb) repr[0]['filesize']\n",
      "6853\n",
      "(Pdb) n\n",
      "> /home/alex/Desktop/urban-physiology/src/glossarizers/ckan-glossarizer/ckan_glossarizer.py(221)write_glossary()\n",
      "-> 'filesize': repr[0]['filesize']\n",
      "(Pdb) \n",
      "> /home/alex/Desktop/urban-physiology/src/glossarizers/ckan-glossarizer/ckan_glossarizer.py(224)write_glossary()\n",
      "-> 'dataset': repr[0]['dataset']\n",
      "(Pdb) \n",
      "> /home/alex/Desktop/urban-physiology/src/glossarizers/ckan-glossarizer/ckan_glossarizer.py(226)write_glossary()\n",
      "-> succeeded = True\n"
     ]
    }
   ],
   "source": [
    "import ckan_glossarizer; ckan_glossarizer.write_glossary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'dataset': '.',\n",
       "  'extension': 'csv',\n",
       "  'filesize': 7622,\n",
       "  'mimetype': 'text/csv',\n",
       "  'resource': 'https://storage.data.gov.sg/air-accident-investigation-reports-by-date/resources/air-accident-investigation-reports-by-date-2016-08-05T02-54-38Z.csv'}]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys; sys.path.append(\"../../limited-requests\"); import limited_requests\n",
    "q = limited_requests.q()\n",
    "limited_requests.limited_get(\"https://storage.data.gov.sg/air-accident-investigation-reports-by-date/resources/air-accident-investigation-reports-by-date-2016-08-05T02-54-38Z.csv\", q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# import sys; sys.path.append(\"../../datafy\")\n",
    "# import datafy\n",
    "# d = datafy.get(\"https://geo.data.gov.sg/afterdeathfacilities/2015/02/02/kml/afterdeathfacilities.zip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Bad mimetype guess: https://geo.data.gov.sg/afterdeathfacilities/2015/02/02/kml/afterdeathfacilities.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# import requests; requests.head(\"https://geo.data.gov.sg/afterdeathfacilities/2015/02/02/kml/afterdeathfacilities.zip\").headers"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
